optimize/local.go

```diff
@@ -4,222 +4,139 @@
 
 package optimize
 
-import (
-	"math"
+import "math"
 
-	"gonum.org/v1/gonum/floats"
-)
+// LocalController controls the optimization process for a LocalMethod.
+// TODO(btracey): Rename?
```

Comment:
Yes, the plan is to make it stay. I was originally intending it would go, but it seems like a nice convenience type to local optimization methods implemented outside the package.

Reply:
Okay, I see the problem. Either the pattern needs to be that a local method initializes itself and then calls `LocalController`, or the pattern should be that `LocalController` calls init on the method. Seems better for the method to be in charge. That probably means there's way too much scaffolding here, I'll try and simplify (maybe just a `RunLocal` method)

Reply:
My original expectation was that by unifying local and global interfaces we would rewrite the reverse communication into a channel communication which feels more Go-like. LocalController translates from reverse to channel communication, allowing the reverse comm design to stay. And this is the reason for my original question because I don't know whether this is good or not.  I would say that removing the reverse communication would simplify things and reduce the API but then having something to help with the common code for local methods would be nice.

Reply:
That was my original thought as well. It's a bit tricky though with the use of the linesearcher. However, with the suggestion of making `LocalController` private, it's easy to move things over piecemeal. One thing I would really love is to have a `RestartLocal` method that performs random-restart descent (in parallel). I believe this is totally compatible with the change, but I have also been keeping that in mind.

Reply:
Ah, there's the `Linesearcher`, I didn't realize.

SGTM.

Reply:
I'm not sure that means we should give up, the problem is just that `Linesearcher` needs to call `NextDirection`, but we could probably switch things around so that the method is in control and not the linesearcher.

Reply:
I would have to look at the code in detail. Anyway, I think that's not in the scope of this PR.

Reply:
`localOptimizer` made private.